apps/consulting/posts/improvements-to-bsr-matrix-multiplication-in-pytorch.md

```diff
@@ -0,0 +1,292 @@
+/>
+
+## Matrix Multiplication with Sparse Tensors
+
+In PyTorch, dense matrix multiplication in eager mode will usually forward to a high performance
+math library (i.e. cuBLAS) that implements the [`GEMM`][cublas-docs-gemm]. Once we start to involve
+sparse tensors the situation becomes more complex. We must have a different `GEMM` like operation
+for every pattern of sparse/dense arguments we need to support. Here we will focus on two forms
+involving exactly one sparse argument, we have
+
+- `C(Dense) += A(Sparse) @ B(Dense)` or **DSD**
+- `C(Sparse) += A(Dense) @ B(Dense)` or **SDD**
+
+where `@` is used to represent the matrix multiplication operation.
+
+The DSD pattern is an operation with the same semantics as a dense matrix multiplication. If the
+sparse argument were converted to dense, or all implicit zeros were materialized the result should
+be equivalent. The DSD pattern can be used to compute linear layers with sparse weights. This is an
+function which is semantically equivalent to [`torch.nn.functional.linear`][pytorch-docs-linear]
+
+```python
+def linear(x, W, b=None):
+    tmp = (W @ x.transpose(-2, -1)).transpose(-2, -1)
+    if b is not None:
+        return tmp + b
+    else:
+        return tmp
+```
+
+In this case the weights, `W` can be stored using BSR layout, provided the DSD matrix multiply is
+supported. Note that linear is defined by `x@W^T` and above we have calculated the transpose of this
+expression.
+
+In contrast the SDD pattern does is not semantically equivalent under conversion to dense layout. In
+this case the sparse tensor updated with the result of `A@B` also acts as a mask.  In other words
+`C` will have the same sparsity pattern and that pattern is imposed on the product `A@B`.  In
+PyTorch, we call this operation [`torch.sparse.sampled_addmm`][pytorch-docs-sampled-addmm].  This
+operation is also relevant for transformer models.  It can be used to compute masked gradients for
+terms like `linear` where during training we want to ensure the sparsity pattern of `W` is preserved.
+
+There exist BLAS-like libraries supporting sparse layouts, for example NVIDIA's
+[cuSPARSE][cusparse-docs]. These options are lacking some important features that are needed to make
+it practical to use sparse layouts in machine learning applications. First, they typically perform
+poorly compared to dense matrix multiply except when the sparsity is very high (<10% of the elements
+are stored). Second, many of the operations are missing support for half precision data types, which
+is a feature used frequently in learning applications.
+
+In order to overcome the limitations, we have explored other options, specifically using the
+[Triton][triton] compiler to author kernels. The `torch.compile` system introduced in PyTorch
+2.0 uses Triton to generate GPU kernels. We now have prototype implementations for `DSD` and `SDD`
+matrix multiplication using BSR layout available in PyTorch 2.1.
```

Comment:
Available in which kernels?

Reply:
Implementation and kernel would be synonymous here. I discuss later where to access the APIs, and how one of these kernels is connected to several operators under particular conditions.

Reply:
Hm. No, I mean... 

> We now have prototype implementations for `DSD` and `SDD` matrix multiplication using BSR layout available in PyTorch 2.1.

`DSD` and `SDD` are algorithms(?), right? And you talk about 'generat[ing] GPU kernels'.

For which kernels/implementations are the `DSD` and `SDD` algorithms now available? All GPUs?

Reply:
The "generation" of kernels is what `torch.compile` does, and for GPUs it will generate triton code. That detail is mentioned to connect triton to what the reader will most likely be familiar with the name from. 

In this case all the kernels are for GPU, triton only supports GPUs, for now. Specifically nvidia GPUs, although some AMD support will be available through the ROCm compatibility layer. I don't think it is necessary to spell all of that out.  We are stating that the kernels use the triton language, and all of our comparisons reference the cuSPARE, and cuBLAS implementations for sparse/dense baselines, so it should be clear that we are talking about GPU kernels.